File: yanyu-jianghu-script/modules/walk.js
/**
 * 地图内行走逻辑（等距视角坐标转像素点击）
 */
const config = require("./config.js");
const mapData = require("./mapData.js");
const Status = require("./status.js");

const Walk = {};

/**
 * 计算当前坐标对应的屏幕像素位置
 */
Walk.calCurrentPixel = function (city, pos) {
    const mapSize = mapData.getMapSize(city);
    const center = config.getCenter();
    const stepX = config.STEP_X * (device.width / config.BASE_WIDTH);
    const stepY = config.STEP_Y * (device.height / config.BASE_HEIGHT);
    const leftOffset = config.LEFT_OFFSET * (device.width / config.BASE_WIDTH);
    const topOffset = config.TOP_OFFSET * (device.height / config.BASE_HEIGHT);

    const sumPixelX = mapSize.width * 0.5 * stepX + mapSize.height * 0.5 * stepX;
    const sumPixelY = mapSize.width * 0.5 * stepY + mapSize.height * 0.5 * stepY;

    const topDisPixel = pos[0] * 0.5 * stepY + pos[1] * 0.5 * stepY;
    const leftDisPixel = pos[0] * 0.5 * stepX + (mapSize.height - pos[1]) * 0.5 * stepX;

    let pixel = [center.x, center.y];

    if (leftDisPixel < center.x) {
        pixel[0] = leftDisPixel + leftOffset;
    } else if (sumPixelX - leftDisPixel < center.x) {
        pixel[0] = device.width - sumPixelX + leftDisPixel;
    } else if (topDisPixel < center.y) {
        pixel[1] = topDisPixel + topOffset;
    } else if (sumPixelY - topDisPixel < center.y) {
        pixel[1] = device.height - sumPixelY + topDisPixel;
    }

    return pixel;
};

/**
 * 点击目标坐标（等距变换）
 */
Walk.goTo = function (currentPixel, fromPos, toPos) {
    const stepX = config.STEP_X * (device.width / config.BASE_WIDTH);
    const stepY = config.STEP_Y * (device.height / config.BASE_HEIGHT);

    const dstX = (toPos[0] - fromPos[0]) * stepX * 0.5 - (toPos[1] - fromPos[1]) * stepX * 0.5;
    const dstY = (toPos[1] - fromPos[1]) * stepY * 0.5 + (toPos[0] - fromPos[0]) * stepY * 0.5;

    const tapX = currentPixel[0] + dstX;
    const tapY = currentPixel[1] + dstY;

    log("点击移动: (" + fromPos[0] + "," + fromPos[1] + ") -> (" + toPos[0] + "," + toPos[1] + ") 像素(" + Math.round(tapX) + "," + Math.round(tapY) + ")");
    click(tapX, tapY);
    sleep(config.DELAY.afterTap);
};

/**
 * 直线行走到目标坐标（分步移动，每步最多 maxStep 格）
 */
Walk.goDirect = function (targetX, targetY, maxStep) {
    maxStep = maxStep || config.MAX_WALK_STEP;
    Status.updatePosition();

    let timeout = 0;
    while (Status.distance(Status.position, [targetX, targetY]) > 0) {
        if (timeout > 30) {
            log("移动超时: 目标(" + targetX + "," + targetY + ")");
            return false;
        }

        const dist = Status.distance(Status.position, [targetX, targetY]);
        const step = Math.min(maxStep, dist);

        // 计算中间点（沿曼哈顿路径）
        let nextX = Status.position[0];
        let nextY = Status.position[1];
        const dx = targetX - Status.position[0];
        const dy = targetY - Status.position[1];

        if (Math.abs(dx) >= Math.abs(dy)) {
            nextX += Math.sign(dx) * Math.min(step, Math.abs(dx));
            const remain = step - Math.min(step, Math.abs(dx));
            nextY += Math.sign(dy) * Math.min(remain, Math.abs(dy));
        } else {
            nextY += Math.sign(dy) * Math.min(step, Math.abs(dy));
            const remain = step - Math.min(step, Math.abs(dy));
            nextX += Math.sign(dx) * Math.min(remain, Math.abs(dx));
        }

        const cp = Walk.calCurrentPixel(Status.city, Status.position);
        Walk.goTo(cp, Status.position, [nextX, nextY]);

        sleep(config.DELAY.afterMove);
        Status.updatePosition();
        timeout++;
    }

    log("到达目标: (" + targetX + "," + targetY + ")");
    return true;
};

/**
 * 行走到指定资源点并尝试采集
 */
Walk.goAndCollect = function (x, y, resourceName) {
    if (!Walk.goDirect(x, y)) return false;

    sleep(config.DELAY.afterMove);

    // 尝试点击采集
    if (Status.clickInteract(config.INTERACT_KEYWORDS.collect)) {
        log("采集: " + resourceName);
        sleep(config.DELAY.afterCollect);
        return true;
    }

    // 备用：点击屏幕中央触发交互
    const center = config.getCenter();
    click(center.x, center.y);
    sleep(config.DELAY.afterTap);

    if (Status.clickInteract(config.INTERACT_KEYWORDS.collect)) {
        log("采集(二次): " + resourceName);
        sleep(config.DELAY.afterCollect);
        return true;
    }

    log("未找到采集选项: " + resourceName);
    return false;
};

module.exports = Walk;

File: yanyu-jianghu-ios/walk.py
"""地图内行走与采集"""

from __future__ import annotations

import logging
import time
from typing import Tuple

import config
import map_data
from status import StatusReader
from wda_client import WdaClient

log = logging.getLogger(__name__)


class Walker:
    def __init__(self, wda: WdaClient, status: StatusReader):
        self.wda = wda
        self.status = status

    def _step_scale(self) -> Tuple[float, float]:
        sx = self.wda.screen_w / config.BASE_WIDTH
        sy = self.wda.screen_h / config.BASE_HEIGHT
        return config.STEP_X * sx, config.STEP_Y * sy

    def cal_current_pixel(self, city: str, pos: Tuple[int, int]) -> Tuple[float, float]:
        mw, mh = map_data.get_map_size(city)
        cx, cy = config.get_center(self.wda.screen_w, self.wda.screen_h)
        step_x, step_y = self._step_scale()
        left = config.LEFT_OFFSET * (self.wda.screen_w / config.BASE_WIDTH)
        top = config.TOP_OFFSET * (self.wda.screen_h / config.BASE_HEIGHT)

        sum_px = mw * 0.5 * step_x + mh * 0.5 * step_x
        sum_py = mw * 0.5 * step_y + mh * 0.5 * step_y
        top_dis = pos[0] * 0.5 * step_y + pos[1] * 0.5 * step_y
        left_dis = pos[0] * 0.5 * step_x + (mh - pos[1]) * 0.5 * step_x

        px, py = cx, cy
        if left_dis < cx:
            px = left_dis + left
        elif sum_px - left_dis < cx:
            px = self.wda.screen_w - sum_px + left_dis
        elif top_dis < cy:
            py = top_dis + top
        elif sum_py - top_dis < cy:
            py = self.wda.screen_h - sum_py + top_dis
        return px, py

    def go_to(self, current_pixel: Tuple[float, float], from_pos: Tuple[int, int], to_pos: Tuple[int, int]) -> None:
        step_x, step_y = self._step_scale()
        dx, dy = to_pos[0] - from_pos[0], to_pos[1] - from_pos[1]
        dst_x = dx * step_x * 0.5 - dy * step_x * 0.5
        dst_y = dy * step_y * 0.5 + dx * step_y * 0.5
        tap_x = current_pixel[0] + dst_x
        tap_y = current_pixel[1] + dst_y
        log.info("移动 (%d,%d)->(%d,%d) tap(%.0f,%.0f)", *from_pos, *to_pos, tap_x, tap_y)
        self.wda.tap(tap_x, tap_y)

    def go_direct(self, target_x: int, target_y: int, max_step: int = config.MAX_WALK_STEP) -> bool:
        self.status.update_position()
        s = self.status.status
        timeout = 0

        while s.distance((target_x, target_y)) > 0:
            if timeout > 30:
                log.error("移动超时 (%d,%d)", target_x, target_y)
                return False

            dist = s.distance((target_x, target_y))
            step = min(max_step, dist)
            nx, ny = s.position
            dx, dy = target_x - nx, target_y - ny

            if abs(dx) >= abs(dy):
                nx += (1 if dx > 0 else -1) * min(step, abs(dx))
                remain = step - min(step, abs(dx))
                ny += (1 if dy > 0 else -1) * min(remain, abs(dy))
            else:
                ny += (1 if dy > 0 else -1) * min(step, abs(dy))
                remain = step - min(step, abs(dy))
                nx += (1 if dx > 0 else -1) * min(remain, abs(dx))

            cp = self.cal_current_pixel(s.city, s.position)
            self.go_to(cp, s.position, (nx, ny))
            time.sleep(config.DELAY["after_move"])
            self.status.update_position()
            timeout += 1

        return True

    def go_and_collect(self, x: int, y: int, resource: str) -> bool:
        if not self.go_direct(x, y):
            return False
        time.sleep(config.DELAY["after_move"])

        if self.status.click_interact(config.INTERACT_KEYWORDS["collect"]):
            log.info("采集: %s", resource)
            time.sleep(config.DELAY["after_collect"])
            return True

        self.wda.tap_center()
        if self.status.click_interact(config.INTERACT_KEYWORDS["collect"]):
            log.info("采集(二次): %s", resource)
            time.sleep(config.DELAY["after_collect"])
            return True

        log.warning("未找到采集选项: %s", resource)
        return False
